From 2eed580ada123c88dcb9670edc209e6c7834d12f Mon Sep 17 00:00:00 2001
From: Yi Xie <imxieyi@users.noreply.github.com>
Date: Tue, 15 Apr 2025 22:23:42 +0900
Subject: [PATCH] Support additional HTTP headers for HTTP playback

Like the cookies stored on a media, the headers stored with
libvlc_media_header_jar_store() are replayed into a jar owned by the
player when libvlc_media_player_play() is called. The http access and
the adaptive demuxer inherit that jar through the "http-headers"
variable. The player jar is cleared on each play so the headers of a
media are not sent for the next one.
---
 include/vlc/libvlc_media.h                    |  30 ++++
 include/vlc_http.h                            |  34 ++++
 include/vlc_player.h                          |  12 ++
 lib/libvlc.sym                                |   2 +
 lib/media.c                                   |  43 +++++
 lib/media_internal.h                          |   5 +
 lib/media_player.c                            |  11 ++
 modules/access/http/access.c                  |   4 +-
 modules/access/http/connmgr.c                 |   9 +-
 modules/access/http/connmgr.h                 |   6 +-
 modules/access/http/message.c                 |  14 ++
 modules/access/http/message.h                 |   3 +
 modules/access/http/outfile.c                 |   1 +
 modules/access/http/resource.c                |   1 +
 modules/access/webdav.c                       |   2 +-
 modules/access_output/http-put.c              |   2 +-
 modules/demux/adaptive/http/AuthStorage.cpp   |   8 +
 modules/demux/adaptive/http/AuthStorage.hpp   |   2 +
 .../demux/adaptive/http/HTTPConnection.cpp    |   6 +-
 src/Makefile.am                               |   1 +
 src/libvlccore.sym                            |   7 +
 src/meson.build                               |   1 +
 src/misc/httpheaders.c                        | 168 ++++++++++++++++++
 src/player/player.c                           |  22 +++
 24 files changed, 386 insertions(+), 8 deletions(-)
 create mode 100644 src/misc/httpheaders.c

diff --git a/include/vlc/libvlc_media.h b/include/vlc/libvlc_media.h
index 1b9eb85acd..417d16c9de 100644
--- a/include/vlc/libvlc_media.h
+++ b/include/vlc/libvlc_media.h
@@ -743,6 +743,36 @@ libvlc_media_cookie_jar_store( libvlc_media_t *p_md, const char *psz_cookie,
 LIBVLC_API void
 libvlc_media_cookie_jar_clear( libvlc_media_t *p_md );
 
+/**
+ * Stores a HTTP header in the header jar.
+ *
+ * \note must be called before the first call of libvlc_media_player_play() to
+ * take effect. The header jar is only used for http/https accesses.
+ *
+ * \version LibVLC 4.0.0 and later.
+ *
+ * \param p_md media descriptor object
+ * \param psz_name header name (must not be NULL)
+ * \param psz_value header value (must not be NULL)
+ *
+ * \return 0 on success, -1 on error.
+ */
+LIBVLC_API int
+libvlc_media_header_jar_store( libvlc_media_t *p_md,
+                               const char *psz_name, const char *psz_value );
+/**
+ * Clear the header jar of a media.
+ *
+ * \note must be called before the first call of libvlc_media_player_play() to
+ * take effect.
+ *
+ * \version LibVLC 4.0.0 and later.
+ *
+ * \param p_md media descriptor object
+ */
+LIBVLC_API void
+libvlc_media_header_jar_clear( libvlc_media_t *p_md );
+
 /** @}*/
 
 # ifdef __cplusplus
diff --git a/include/vlc_http.h b/include/vlc_http.h
index 976f6e664a..70a66153a3 100644
--- a/include/vlc_http.h
+++ b/include/vlc_http.h
@@ -98,4 +98,38 @@ VLC_API bool vlc_http_cookies_store( vlc_http_cookie_jar_t *jar,
 VLC_API char *vlc_http_cookies_fetch( vlc_http_cookie_jar_t *jar, bool secure,
                                       const char *host, const char *path );
 
+/* HTTP headers */
+
+typedef struct vlc_http_header_jar_t vlc_http_header_jar_t;
+
+typedef int (*vlc_http_header_foreach_func) (void*, const char*, const char*);
+
+VLC_API vlc_http_header_jar_t * vlc_http_headers_new( void ) VLC_USED;
+VLC_API vlc_http_header_jar_t * vlc_http_headers_copy( vlc_http_header_jar_t * old) VLC_USED;
+VLC_API void vlc_http_headers_clear( vlc_http_header_jar_t * p_jar );
+VLC_API void vlc_http_headers_destroy( vlc_http_header_jar_t * p_jar );
+
+/**
+ * Parse a value of an incoming Set-Cookie header and append the
+ * header to the header jar if appropriate.
+ *
+ * @param jar header jar object
+ * @param name name of the header
+ * @param value value of the header
+ * @return true, if the header was added, false otherwise
+ */
+VLC_API bool vlc_http_headers_store( vlc_http_header_jar_t *jar,
+                                     const char *name, const char *value);
+
+/**
+ * Runs a function on all headers. Stops and returns when non-zero code
+ * is returned from the function.
+ *
+ * @param jar a header jar
+ * @param func the function to run
+ * @param obj the object to pass to the function
+ * @return the return value from the function, if not zero
+ */
+VLC_API int vlc_http_headers_foreach( vlc_http_header_jar_t *jar, vlc_http_header_foreach_func func, void* obj );
+
 #endif /* VLC_HTTP_H */
diff --git a/include/vlc_player.h b/include/vlc_player.h
index f21e763302..6a24c62dea 100644
--- a/include/vlc_player.h
+++ b/include/vlc_player.h
@@ -126,6 +126,18 @@ vlc_player_Delete(vlc_player_t *player);
 VLC_API struct vlc_http_cookie_jar_t *
 vlc_player_GetCookieJar(vlc_player_t *player);
 
+/**
+ * Get the HTTP header jar shared by the player
+ *
+ * This is the jar inherited by the http access of any media played by this
+ * player. It is owned by the player and valid until vlc_player_Delete().
+ *
+ * @param player player instance created by vlc_player_New()
+ * @return the header jar, or NULL if the player has none
+ */
+VLC_API struct vlc_http_header_jar_t *
+vlc_player_GetHeaderJar(vlc_player_t *player);
+
 /**
  * Lock the player.
  *
diff --git a/lib/libvlc.sym b/lib/libvlc.sym
index 7b702ca6a7..c533f554ac 100644
--- a/lib/libvlc.sym
+++ b/lib/libvlc.sym
@@ -223,6 +223,8 @@ libvlc_media_get_meta_extra_names
 libvlc_media_meta_extra_names_release
 libvlc_media_cookie_jar_store
 libvlc_media_cookie_jar_clear
+libvlc_media_header_jar_store
+libvlc_media_header_jar_clear
 libvlc_new
 libvlc_release
 libvlc_renderer_item_name
diff --git a/lib/media.c b/lib/media.c
index acc8570f69..ee9f82cc44 100644
--- a/lib/media.c
+++ b/lib/media.c
@@ -229,6 +229,9 @@ static void media_destroy( void *libvlc_owner )
 
     media_cookies_clear( p_md );
 
+    if( p_md->p_header_jar )
+        vlc_http_headers_destroy( p_md->p_header_jar );
+
     free( p_md );
 }
 
@@ -274,6 +277,7 @@ libvlc_media_t * libvlc_media_new_from_input_item(input_item_t *p_input_item )
     p_md->req = NULL;
     p_md->p_cookies = NULL;
     p_md->i_cookies = 0;
+    p_md->p_header_jar = NULL;
 
     input_item_Hold( p_md->p_input_item );
 
@@ -887,3 +891,42 @@ libvlc_media_cookies_replay( libvlc_media_t *p_md,
                                 p_md->p_cookies[i].psz_host,
                                 p_md->p_cookies[i].psz_path );
 }
+
+int
+libvlc_media_header_jar_store( libvlc_media_t *p_md,
+                               const char *psz_name, const char *psz_value )
+{
+    if( !p_md->p_header_jar )
+    {
+        p_md->p_header_jar = vlc_http_headers_new();
+        if( !p_md->p_header_jar )
+            return -1;
+    }
+    return vlc_http_headers_store( p_md->p_header_jar, psz_name, psz_value ) ? 0 : -1;
+}
+
+void
+libvlc_media_header_jar_clear( libvlc_media_t *p_md )
+{
+    if( p_md->p_header_jar )
+        vlc_http_headers_clear( p_md->p_header_jar );
+}
+
+static int
+media_header_replay_cb( void *opaque, const char *psz_name,
+                        const char *psz_value )
+{
+    vlc_http_header_jar_t *p_jar = opaque;
+    vlc_http_headers_store( p_jar, psz_name, psz_value );
+    return 0;
+}
+
+void
+libvlc_media_headers_replay( libvlc_media_t *p_md,
+                             vlc_http_header_jar_t *p_jar )
+{
+    vlc_http_headers_clear( p_jar );
+    if( p_md->p_header_jar )
+        vlc_http_headers_foreach( p_md->p_header_jar,
+                                  media_header_replay_cb, p_jar );
+}
diff --git a/lib/media_internal.h b/lib/media_internal.h
index e3ba8f254c..b6c2fd99ec 100644
--- a/lib/media_internal.h
+++ b/lib/media_internal.h
@@ -67,12 +67,17 @@ struct libvlc_media_t
 
     struct libvlc_media_cookie_t *p_cookies;
     size_t i_cookies;
+    vlc_http_header_jar_t *p_header_jar;
 };
 
 /* Replay the cookies stored on p_md into p_jar. */
 void libvlc_media_cookies_replay( libvlc_media_t *p_md,
                                   vlc_http_cookie_jar_t *p_jar );
 
+/* Replace the content of p_jar with the headers stored on p_md. */
+void libvlc_media_headers_replay( libvlc_media_t *p_md,
+                                  vlc_http_header_jar_t *p_jar );
+
 /* Media Descriptor */
 libvlc_media_t * libvlc_media_new_from_input_item( input_item_t * );
 
diff --git a/lib/media_player.c b/lib/media_player.c
index 3fffa9145a..98bcc8c187 100644
--- a/lib/media_player.c
+++ b/lib/media_player.c
@@ -1056,6 +1056,17 @@ int libvlc_media_player_play( libvlc_media_player_t *p_mi )
 
         if( p_md != NULL && p_jar != NULL )
             libvlc_media_cookies_replay( p_md, p_jar );
+
+        /* Same for the additional HTTP headers, which replace the ones of
+         * the previously played media. */
+        vlc_http_header_jar_t *p_header_jar = vlc_player_GetHeaderJar( player );
+        if( p_header_jar != NULL )
+        {
+            if( p_md != NULL )
+                libvlc_media_headers_replay( p_md, p_header_jar );
+            else
+                vlc_http_headers_clear( p_header_jar );
+        }
     }
 
     int ret = vlc_player_Start(player);
diff --git a/modules/access/http/access.c b/modules/access/http/access.c
index ffc38b749e..f775a1f1a6 100644
--- a/modules/access/http/access.c
+++ b/modules/access/http/access.c
@@ -168,6 +168,8 @@ static int Open(vlc_object_t *obj)
     if (var_InheritBool(obj, "http-forward-cookies"))
         jar = var_InheritAddress(obj, "http-cookies");
 
+    void *header_jar = var_InheritAddress(obj, "http-headers");
+
     struct vlc_credential crd;
     struct vlc_url_t crd_url;
     char *psz_realm = NULL;
@@ -175,7 +177,7 @@ static int Open(vlc_object_t *obj)
     vlc_UrlParse(&crd_url, access->psz_url);
     vlc_credential_init(&crd, &crd_url);
 
-    sys->manager = vlc_http_mgr_create(obj, jar);
+    sys->manager = vlc_http_mgr_create(obj, jar, header_jar);
     if (sys->manager == NULL)
         goto error;
 
diff --git a/modules/access/http/connmgr.c b/modules/access/http/connmgr.c
index ca734853fc..4eedd6a4eb 100644
--- a/modules/access/http/connmgr.c
+++ b/modules/access/http/connmgr.c
@@ -102,6 +102,7 @@ struct vlc_http_mgr
     vlc_object_t *obj;
     vlc_tls_client_t *creds;
     struct vlc_http_cookie_jar_t *jar;
+    struct vlc_http_header_jar_t *header_jar;
     struct vlc_http_conn *conn;
 };
 
@@ -287,8 +288,13 @@ struct vlc_http_cookie_jar_t *vlc_http_mgr_get_jar(struct vlc_http_mgr *mgr)
     return mgr->jar;
 }
 
+struct vlc_http_header_jar_t *vlc_http_mgr_get_header_jar(struct vlc_http_mgr *mgr) {
+    return mgr->header_jar;
+}
+
 struct vlc_http_mgr *vlc_http_mgr_create(vlc_object_t *obj,
-                                         struct vlc_http_cookie_jar_t *jar)
+                                         struct vlc_http_cookie_jar_t *jar,
+                                         struct vlc_http_header_jar_t *header_jar)
 {
     struct vlc_http_mgr *mgr = malloc(sizeof (*mgr));
     if (unlikely(mgr == NULL))
@@ -298,6 +304,7 @@ struct vlc_http_mgr *vlc_http_mgr_create(vlc_object_t *obj,
     mgr->obj = obj;
     mgr->creds = NULL;
     mgr->jar = jar;
+    mgr->header_jar = header_jar;
     mgr->conn = NULL;
     return mgr;
 }
diff --git a/modules/access/http/connmgr.h b/modules/access/http/connmgr.h
index b7faa898af..78a10c8dd6 100644
--- a/modules/access/http/connmgr.h
+++ b/modules/access/http/connmgr.h
@@ -32,6 +32,7 @@
 struct vlc_http_mgr;
 struct vlc_http_msg;
 struct vlc_http_cookie_jar_t;
+struct vlc_http_header_jar_t;
 
 /**
  * Sends an HTTP request
@@ -56,6 +57,7 @@ struct vlc_http_msg *vlc_http_mgr_request(struct vlc_http_mgr *mgr, bool https,
                                           bool idempotent, bool payload);
 
 struct vlc_http_cookie_jar_t *vlc_http_mgr_get_jar(struct vlc_http_mgr *);
+struct vlc_http_header_jar_t *vlc_http_mgr_get_header_jar(struct vlc_http_mgr *);
 
 /**
  * Creates an HTTP connection manager
@@ -64,9 +66,11 @@ struct vlc_http_cookie_jar_t *vlc_http_mgr_get_jar(struct vlc_http_mgr *);
  *
  * @param obj parent VLC object
  * @param jar HTTP cookies jar (NULL to disable cookies)
+ * @param header_jar HTTP headers jar (NULL to disable additional headers)
  */
 struct vlc_http_mgr *vlc_http_mgr_create(vlc_object_t *obj,
-                                         struct vlc_http_cookie_jar_t *jar);
+                                         struct vlc_http_cookie_jar_t *jar,
+                                         struct vlc_http_header_jar_t *header_jar);
 
 /**
  * Destroys an HTTP connection manager
diff --git a/modules/access/http/message.c b/modules/access/http/message.c
index ba1a3f632c..fab8bc4425 100644
--- a/modules/access/http/message.c
+++ b/modules/access/http/message.c
@@ -984,6 +984,20 @@ int vlc_http_msg_add_cookies(struct vlc_http_msg *m,
     return val;
 }
 
+int vlc_http_add_header_internal(void *obj, const char* name, const char* value) {
+    struct vlc_http_msg *m = obj;
+    return vlc_http_msg_add_header(m, name, "%s", value);
+}
+
+int vlc_http_msg_add_headers(struct vlc_http_msg *m,
+                             vlc_http_header_jar_t *jar)
+{
+    if (jar == NULL)
+        return 0;
+
+    return vlc_http_headers_foreach(jar, &vlc_http_add_header_internal, m);
+}
+
 char *vlc_http_msg_get_basic_realm(const struct vlc_http_msg *m)
 {
     const char *auth;
diff --git a/modules/access/http/message.h b/modules/access/http/message.h
index 9abfd131b3..68f9668c55 100644
--- a/modules/access/http/message.h
+++ b/modules/access/http/message.h
@@ -30,6 +30,7 @@
 
 struct vlc_http_msg;
 struct vlc_http_cookie_jar_t;
+struct vlc_http_header_jar_t;
 
 /**
  * Creates an HTTP request.
@@ -169,6 +170,8 @@ void vlc_http_msg_get_cookies(const struct vlc_http_msg *,
                               const char *host, const char *path);
 int vlc_http_msg_add_cookies(struct vlc_http_msg *,
                              struct vlc_http_cookie_jar_t *);
+int vlc_http_msg_add_headers(struct vlc_http_msg *,
+                             struct vlc_http_header_jar_t *);
 
 char *vlc_http_msg_get_basic_realm(const struct vlc_http_msg *);
 
diff --git a/modules/access/http/outfile.c b/modules/access/http/outfile.c
index d31c825a2c..5444e00eb3 100644
--- a/modules/access/http/outfile.c
+++ b/modules/access/http/outfile.c
@@ -87,6 +87,7 @@ struct vlc_http_outfile *vlc_http_outfile_create(struct vlc_http_mgr *mgr,
         vlc_http_msg_add_agent(req, ua);
 
     vlc_http_msg_add_cookies(req, vlc_http_mgr_get_jar(mgr));
+    vlc_http_msg_add_headers(req, vlc_http_mgr_get_header_jar(mgr));
 
     resp = vlc_http_mgr_request(mgr, secure, url.psz_host, url.i_port, req,
                                 false, true);
diff --git a/modules/access/http/resource.c b/modules/access/http/resource.c
index d15b1decc4..c626ecaf93 100644
--- a/modules/access/http/resource.c
+++ b/modules/access/http/resource.c
@@ -71,6 +71,7 @@ vlc_http_res_req(const struct vlc_http_resource *res, void *opaque)
         vlc_http_msg_add_header(req, "Referer", "%s", res->referrer);
 
     vlc_http_msg_add_cookies(req, vlc_http_mgr_get_jar(res->manager));
+    vlc_http_msg_add_headers(req, vlc_http_mgr_get_header_jar(res->manager));
 
     /* TODO: vlc_http_msg_add_header(req, "TE", "gzip, deflate"); */
 
diff --git a/modules/access/webdav.c b/modules/access/webdav.c
index a744bf43c9..c5ce3dd524 100644
--- a/modules/access/webdav.c
+++ b/modules/access/webdav.c
@@ -525,7 +525,7 @@ static int Open(vlc_object_t *obj)
     if (sys->path == NULL)
         goto clean;
 
-    sys->manager = vlc_http_mgr_create(obj, NULL);
+    sys->manager = vlc_http_mgr_create(obj, NULL, NULL);
     if (sys->manager == NULL)
         goto clean;
 
diff --git a/modules/access_output/http-put.c b/modules/access_output/http-put.c
index 21d4af3a6a..9548725001 100644
--- a/modules/access_output/http-put.c
+++ b/modules/access_output/http-put.c
@@ -78,7 +78,7 @@ static int Open(vlc_object_t *obj)
     if (unlikely(sys == NULL))
         return VLC_ENOMEM;
 
-    sys->manager = vlc_http_mgr_create(obj, NULL);
+    sys->manager = vlc_http_mgr_create(obj, NULL, NULL);
     if (sys->manager == NULL)
         return VLC_ENOMEM;
 
diff --git a/modules/demux/adaptive/http/AuthStorage.cpp b/modules/demux/adaptive/http/AuthStorage.cpp
index 0e4458b580..f7ce2f07e0 100644
--- a/modules/demux/adaptive/http/AuthStorage.cpp
+++ b/modules/demux/adaptive/http/AuthStorage.cpp
@@ -33,6 +33,9 @@ AuthStorage::AuthStorage( vlc_object_t *p_obj )
                 (var_InheritAddress( p_obj, "http-cookies" ));
     else
         p_cookies_jar = nullptr;
+    
+    p_headers_jar = static_cast<vlc_http_header_jar_t *>
+        (var_InheritAddress( p_obj, "http-headers" ));
 }
 
 AuthStorage::~AuthStorage()
@@ -67,3 +70,8 @@ vlc_http_cookie_jar_t *AuthStorage::getJar() const
 {
     return p_cookies_jar;
 }
+
+vlc_http_header_jar_t *AuthStorage::getHeadersJar() const
+{
+    return p_headers_jar;
+}
diff --git a/modules/demux/adaptive/http/AuthStorage.hpp b/modules/demux/adaptive/http/AuthStorage.hpp
index 20c791dec0..5a1e5ec6f6 100644
--- a/modules/demux/adaptive/http/AuthStorage.hpp
+++ b/modules/demux/adaptive/http/AuthStorage.hpp
@@ -39,9 +39,11 @@ namespace adaptive
                 void addCookie( const std::string &cookie, const ConnectionParams & );
                 std::string getCookie( const ConnectionParams &, bool secure );
                 vlc_http_cookie_jar_t *getJar() const;
+                vlc_http_header_jar_t *getHeadersJar() const;
 
             private:
                 vlc_http_cookie_jar_t *p_cookies_jar;
+                vlc_http_header_jar_t *p_headers_jar;
         };
     }
 }
diff --git a/modules/demux/adaptive/http/HTTPConnection.cpp b/modules/demux/adaptive/http/HTTPConnection.cpp
index deb1c2f934..0192bcd5d2 100644
--- a/modules/demux/adaptive/http/HTTPConnection.cpp
+++ b/modules/demux/adaptive/http/HTTPConnection.cpp
@@ -89,10 +89,10 @@ const ConnectionParams & AbstractConnection::getRedirection() const
 class adaptive::http::LibVLCHTTPSource : public adaptive::BlockStreamInterface
 {
      public:
-        LibVLCHTTPSource(vlc_object_t *p_object_, struct vlc_http_cookie_jar_t *jar)
+        LibVLCHTTPSource(vlc_object_t *p_object_, struct vlc_http_cookie_jar_t *jar, struct vlc_http_header_jar_t *header_jar)
         {
             p_object = p_object_;
-            http_mgr = vlc_http_mgr_create(p_object, jar);
+            http_mgr = vlc_http_mgr_create(p_object, jar, header_jar);
             http_res = nullptr;
             totalRead = 0;
         }
@@ -352,7 +352,7 @@ const struct vlc_http_resource_cbs LibVLCHTTPSource::callbacks =
 LibVLCHTTPConnection::LibVLCHTTPConnection(vlc_object_t *p_object_, AuthStorage *auth)
     : AbstractConnection( p_object_ )
 {
-    source = new adaptive::http::LibVLCHTTPSource(p_object_, auth->getJar());
+    source = new adaptive::http::LibVLCHTTPSource(p_object_, auth->getJar(), auth->getHeadersJar());
     sourceStream = new ChunksSourceStream(p_object, source);
     stream = nullptr;
     char *psz_useragent = var_InheritString(p_object_, "http-user-agent");
diff --git a/src/Makefile.am b/src/Makefile.am
index 9d93ba4df5..610c264398 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -434,6 +434,7 @@ libvlccore_la_SOURCES = \
 	misc/filter.c \
 	misc/filter_chain.c \
 	misc/httpcookies.c \
+	misc/httpheaders.c \
 	misc/fingerprinter.c \
 	misc/text_style.c \
 	misc/sort.c \
diff --git a/src/libvlccore.sym b/src/libvlccore.sym
index 258eab9e12..010a7ef6c8 100644
--- a/src/libvlccore.sym
+++ b/src/libvlccore.sym
@@ -154,6 +154,12 @@ vlc_http_cookies_destroy
 vlc_http_cookies_clear
 vlc_http_cookies_store
 vlc_http_cookies_fetch
+vlc_http_headers_new
+vlc_http_headers_copy
+vlc_http_headers_clear
+vlc_http_headers_destroy
+vlc_http_headers_store
+vlc_http_headers_foreach
 httpd_ClientIP
 httpd_FileDelete
 httpd_FileNew
@@ -872,6 +878,7 @@ vlc_player_GetCapabilities
 vlc_player_GetCategoryDelay
 vlc_player_GetCategoryLanguage
 vlc_player_GetCookieJar
+vlc_player_GetHeaderJar
 vlc_player_GetCurrentMedia
 vlc_player_GetNextMedia
 vlc_player_GetError
diff --git a/src/meson.build b/src/meson.build
index d142f56186..50c93a836d 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -294,6 +294,7 @@ libvlccore_sources_base = files(
     'misc/filter.c',
     'misc/filter_chain.c',
     'misc/httpcookies.c',
+    'misc/httpheaders.c',
     'misc/fingerprinter.c',
     'misc/text_style.c',
     'misc/sort.c',
diff --git a/src/misc/httpheaders.c b/src/misc/httpheaders.c
new file mode 100644
index 0000000000..490735fb5b
--- /dev/null
+++ b/src/misc/httpheaders.c
@@ -0,0 +1,168 @@
+/*****************************************************************************
+ * httpheaders.c: HTTP headers utilities
+ *****************************************************************************
+ * Copyright (C) 2025 VLC authors and VideoLAN
+ *
+ * Authors: Yi Xie <imxieyi@github>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+#include <vlc_common.h>
+#include <vlc_arrays.h>
+#include <vlc_messages.h>
+#include <vlc_strings.h>
+#include <vlc_http.h>
+#include <vlc_threads.h>
+
+typedef struct http_header_t
+{
+    char *psz_name;
+    char *psz_value;
+} http_header_t;
+
+static void header_destroy(http_header_t *header)
+{
+    assert(header != NULL);
+    free(header->psz_name);
+    free(header->psz_value);
+    free(header);
+}
+
+VLC_MALLOC VLC_USED
+static http_header_t * header_copy(http_header_t * old) {
+    http_header_t* new = malloc(sizeof(http_header_t));
+    new->psz_name = strdup(old->psz_name);
+    new->psz_value = strdup(old->psz_value);
+    return new;
+}
+
+struct vlc_http_header_jar_t
+{
+    vlc_array_t headers;
+    vlc_mutex_t lock;
+};
+
+vlc_http_header_jar_t * vlc_http_headers_new(void)
+{
+    vlc_http_header_jar_t * jar = malloc( sizeof( vlc_http_header_jar_t ) );
+    if ( unlikely(jar == NULL) )
+        return NULL;
+
+    vlc_array_init( &jar->headers );
+    vlc_mutex_init( &jar->lock );
+
+    return jar;
+}
+
+vlc_http_header_jar_t * vlc_http_headers_copy( vlc_http_header_jar_t * old)
+{
+    vlc_http_header_jar_t* copy = vlc_http_headers_new();
+    for( size_t i = 0; i < vlc_array_count( &old->headers ); i++ )
+        vlc_array_append(&copy->headers, header_copy(vlc_array_item_at_index(&old->headers, i)));
+    return copy;
+}
+
+void vlc_http_headers_clear( vlc_http_header_jar_t * p_jar )
+{
+    if ( !p_jar )
+        return;
+
+    vlc_mutex_lock( &p_jar->lock );
+
+    for( size_t i = 0; i < vlc_array_count( &p_jar->headers ); i++ )
+        header_destroy( vlc_array_item_at_index( &p_jar->headers, i ) );
+
+    vlc_array_clear( &p_jar->headers );
+
+    vlc_mutex_unlock( &p_jar->lock );
+}
+
+void vlc_http_headers_destroy( vlc_http_header_jar_t * p_jar )
+{
+    vlc_http_headers_clear( p_jar );
+    free( p_jar );
+}
+
+bool vlc_http_headers_store(vlc_http_header_jar_t *p_jar, const char *name,
+                            const char *value)
+{
+    assert(name != NULL);
+    assert(value != NULL);
+
+    vlc_mutex_lock( &p_jar->lock );
+
+    for( size_t i = 0; i < vlc_array_count( &p_jar->headers ); i++ )
+    {
+        http_header_t *iter = vlc_array_item_at_index( &p_jar->headers, i );
+
+        assert( iter->psz_name );
+        assert( iter->psz_value );
+
+        if( vlc_ascii_strcasecmp( name, iter->psz_name ) == 0 )
+        {
+            /* Remove previous value for this header */
+            vlc_array_remove( &p_jar->headers, i );
+            header_destroy(iter);
+            break;
+        }
+    }
+    http_header_t *header = malloc(sizeof (*header));
+    if (unlikely(header == NULL))
+    {
+        vlc_mutex_unlock( &p_jar->lock );
+        return false;
+    }
+
+    header->psz_name = strdup(name);
+    header->psz_value = strdup(value);
+
+    bool b_ret = (vlc_array_append( &p_jar->headers, header ) == 0);
+    if( !b_ret )
+        header_destroy( header );
+
+    vlc_mutex_unlock( &p_jar->lock );
+
+    return b_ret;
+}
+
+typedef int (*vlc_http_header_foreach_func) (void*, const char*, const char*);
+
+int vlc_http_headers_foreach(vlc_http_header_jar_t *p_jar, vlc_http_header_foreach_func func, void* obj)
+{
+    vlc_mutex_lock( &p_jar->lock );
+
+    int val = 0;
+    http_header_t* header = NULL;
+    for (int i = 0; i < p_jar->headers.i_count; i++) {
+        header = vlc_array_item_at_index(&p_jar->headers, i);
+        val = (*func)(obj, header->psz_name, header->psz_value);
+        if (val) {
+            break;
+        }
+    }
+
+    vlc_mutex_unlock( &p_jar->lock );
+    return val;
+}
diff --git a/src/player/player.c b/src/player/player.c
index 6f9b33921f..41f1fbccaa 100644
--- a/src/player/player.c
+++ b/src/player/player.c
@@ -2091,6 +2091,13 @@ vlc_player_Delete(vlc_player_t *player)
         vlc_http_cookies_destroy(cookies);
     }
 
+    vlc_http_header_jar_t *headers = var_GetAddress(player, "http-headers");
+    if (headers != NULL)
+    {
+        var_Destroy(player, "http-headers");
+        vlc_http_headers_destroy(headers);
+    }
+
     assert(!vlc_mutex_held(&player->lock));
 
     vlc_object_delete(player);
@@ -2102,6 +2109,12 @@ vlc_player_GetCookieJar(vlc_player_t *player)
     return var_GetAddress(player, "http-cookies");
 }
 
+vlc_http_header_jar_t *
+vlc_player_GetHeaderJar(vlc_player_t *player)
+{
+    return var_GetAddress(player, "http-headers");
+}
+
 vlc_player_t *
 vlc_player_New(vlc_object_t *parent, enum vlc_player_lock_type lock_type)
 {
@@ -2177,6 +2190,15 @@ vlc_player_New(vlc_object_t *parent, enum vlc_player_lock_type lock_type)
         VAR_CREATE("http-cookies", VLC_VAR_ADDRESS);
         var_SetChecked(player, "http-cookies", VLC_VAR_ADDRESS, cookies);
     }
+
+    /* Initialize the shared HTTP header jar */
+    vlc_value_t headers;
+    headers.p_address = vlc_http_headers_new();
+    if (likely(headers.p_address != NULL))
+    {
+        VAR_CREATE("http-headers", VLC_VAR_ADDRESS);
+        var_SetChecked(player, "http-headers", VLC_VAR_ADDRESS, headers);
+    }
 #undef VAR_CREATE
 
     player->resource = input_resource_New(VLC_OBJECT(player));
-- 
2.54.0 (Apple Git-157)

